### PR TITLE
[Chat][Feature] messages fetching to manage network/connect/disconnect scenario

### PR DESCRIPTION
**Pull request by iaulakh**
## Purpose

* polling implementation to manage network disconnect

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Pull Request Type
What kind of change does this Pull Request introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

### CODE REVIEW

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPollingHandler.kt [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        }
+    }
+
+    /**
+     * RingBuffer uses a fixed length array to implement a queue, where,
+     * - [tail] Items are added to the tail
+     * - [head] Items are removed from the head
+     * - [capacity] Keeps track of how many items are currently in the queue
+     */
+    internal class RingBuffer<T>(private val maxSize: Int = 0) {
```
Comment (ahammer) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPollingHandler.kt:252]: Can Deque be used instead?
https://docs.oracle.com/javase/7/docs/api/java/util/Deque.html

Comment (iaulakh) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPollingHandler.kt:216]: Adam, it seems like initialCapacity?

Comment (iaulakh) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPollingHandler.kt:216]: I think it accepts initial limit that increases itself?

Comment (iaulakh) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPollingHandler.kt]: removed as per today discussion.

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/presentation/manager/NetworkManager.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import android.net.Network
+import android.net.NetworkCapabilities
+import android.net.NetworkRequest
+import com.azure.android.communication.ui.chat.redux.Dispatch
+import com.azure.android.communication.ui.chat.redux.action.NetworkAction
+
+internal class NetworkManager(dispatch: Dispatch) {
+    private var connectivityManager: ConnectivityManager? = null
+
+    fun start(context: Context) {
```
Comment (ahammer) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/presentation/manager/NetworkManager.kt:17]: needs cleanup/stop methods

Comment (ahammer) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/presentation/manager/NetworkManager.kt:17]: https://developer.android.com/reference/android/net/ConnectivityManager#unregisterNetworkCallback(android.net.ConnectivityManager.NetworkCallback)

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/action/NetworkAction.kt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+// Licensed under the MIT License.
+
+package com.azure.android.communication.ui.chat.redux.action
+
+import org.threeten.bp.OffsetDateTime
+
+internal sealed class NetworkAction : Action {
+    class Connected : NetworkAction()
+    class Disconnected : NetworkAction()
+    class SetDisconnectedOffset(val disconnectOffsetDateTime: OffsetDateTime) : NetworkAction()
```
Comment (ahammer) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/action/NetworkAction.kt:11]: Can Time be Implicit or pass the disconnected time in the above action?

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/repository/RepositoryMiddleware.kt [hunk 4]
```diff
@@ -35,13 +36,26 @@ internal class RepositoryMiddlewareImpl(
...
                 is ChatAction.MessageEdited -> processEditMessage(action, store::dispatch)
+                is NetworkAction.Disconnected -> processNetworkDisconnected(store::dispatch)
             }
 
             // Pass Action down the chain
             next(action)
         }
     }
 
+    private fun processNetworkDisconnected(
```
Comment (ahammer) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/repository/RepositoryMiddleware.kt:47]: Not sure this belongs in this class, as it's not instrumenting MessageRepository. (this should just be installing/removing/editing/deleting messages in the repo)

Maybe we need NetworkMiddleware?

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatActionHandler.kt [hunk 5]
```diff
@@ -6,15 +6,31 @@ package com.azure.android.communication.ui.chat.redux.middleware.sdk
...
+import com.azure.android.communication.ui.chat.redux.action.NetworkAction
+import com.azure.android.communication.ui.chat.redux.state.NetworkStatus
+import com.azure.android.communication.ui.chat.redux.state.ReduxState
 import com.azure.android.communication.ui.chat.service.ChatService
 
 // Converts Redux Actions into SDK Calls
 // Redux -> Service
 internal class ChatActionHandler(private val chatService: ChatService) {
 
+    fun onAction(store: Store<ReduxState>, action: Action) {
```
Comment (ahammer) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatActionHandler.kt:22]: add ReduxState to the other onAction so we don't have 2 onActions in the ChatActionHandler

However, feel this action should be handled by a separate middleware that brings in 
NetworkAction.Connected and Emits a ChatAction that results in FetchMessages call.

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPullEventHandler.kt [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+import com.azure.android.core.util.RequestContext
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import kotlinx.coroutines.asCoroutineDispatcher
+import org.threeten.bp.OffsetDateTime
+import java.util.concurrent.Executors
+
+internal class ChatPullEventHandler(coroutineContextProvider: CoroutineContextProvider) {
```
Comment (ahammer) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPullEventHandler.kt:26]: Pull naming could be confusing.

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPullEventHandler.kt [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        this.chatThreadClient = chatThreadClient
+        this.eventSubscriber = eventSubscriber
+    }
+
+    fun stop() {
+        singleThreadedContext.shutdown()
+        coroutineScope.cancel()
+    }
+
+    private fun fetchMessagesList(from: OffsetDateTime?) {
```
Comment (ahammer) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPullEventHandler.kt:59]: Long method, if possible refactor to functions for readability.

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/action/NetworkAction.kt (hunk 3 above)
Comment (iaulakh) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/action/NetworkAction.kt:11]: discussed, no action required

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPullEventHandler.kt (hunk 7 above)
Comment (iaulakh) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPullEventHandler.kt]: resolving as per discussion

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/repository/RepositoryMiddleware.kt (hunk 4 above)
Comment (iaulakh) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/repository/RepositoryMiddleware.kt:47]: resolving as per discussion

**Review by ShaunaSong — DISMISSED** [review]

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/presentation/manager/NetworkManager.kt [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
+            .build()
+        connectivityManager =
+            context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+        connectivityManager?.requestNetwork(networkRequest, networkCallback)
+    }
+
+    fun stop() {
+        connectivityManager?.unregisterNetworkCallback(networkCallback)
+    }
```
Comment (Mohtasim) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/presentation/manager/NetworkManager.kt:30]: Can we have tests on this manager class?

Comment (iaulakh) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/presentation/manager/NetworkManager.kt:30]: It has UI logic Mohtasim. We skip for UI layer

connectivityManager =
            context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

**Review by ShaunaSong — APPROVED** [review]

**Review by Mohtasim — APPROVED** [review]

**Pull request merged**